### ISSUE TITLE
add container FQDN to /etc/hosts

### ISSUE THREAD

**Issue by rmharris157** [weaveworks/weave#1497]
Is there a way to get weaveDNS to write a containers /etc/hosts file using the container's hostname's FQDN? Not doing so is causing pain in Ambari which wants FQDN's for everything.

**Comment by rade** [comment]
So just to be clear, instead of `/etc/hosts` containing an entry with just the containers hostname (as happens now), you want that entry to include the domain?

That really should not make any difference to any programs that use standard resolver libraries and API functions. For example, `hostname` does return the fully qualified name already.

Does Ambari do its own (broken) processing of `/etc/hosts`?

**Comment by rmharris157** [comment]
Yes—it seems that Ambari’s remote client does a reverse lookup after the initial, which it then pulls from the container’s local /etc/hosts.

However, I did some digging last night—let me rephrase the question.

Let’s say I reboot my CoreOS box with all the defaults. I ssh in and start weave and a basic container and then look at the /etc/hosts generated by Weave.

rharris@febtober> ssh core@192.168.1.239
Last login: Fri Oct  2 10:55:03 2015 from 10.8.0.10
CoreOS stable (766.3.0)
core@alpha ~ $ uptime
 10:55:53 up 0 min,  1 user,  load average: 0.44, 0.10, 0.03
core@alpha ~ $ docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
core@alpha ~ $ ps -aef | grep weave
core       836   700  0 10:57 pts/0    00:00:00 grep --colour=auto weave
core@alpha ~ $ weave launch
Setting docker0 MAC (mitigate https://github.com/docker/docker/issues/14908)
core@alpha ~ $ eval $(weave env)
core@alpha ~ $ docker run -it --name test1 -h test1.weave.local centos bash
[root@test /]# cat /etc/hosts
# modified by weave
10.32.0.1	test1
127.0.0.1	localhost
::1	ip6-localhost ip6-loopback localhost
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters

Boo. It didn’t create /etc/hosts the way I wanted. Let me spin up another container in another window, perhaps to try a ping test.

rharris@febtober> ssh core@192.168.1.239
Last login: Fri Oct  2 10:55:50 2015 from 10.8.0.10
CoreOS stable (766.3.0)
core@alpha ~ $ docker run -it --name test2 -h test2.weave.local centos bash
[root@test2 /]# cat /etc/hosts
172.17.0.4	test2.weave.local test2 
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters

Wait-it just did it?!?! What that…

Also notice something else here. I didn’t set the environment before starting the container. Why does that control how /etc/hosts gets written back to the container?

Further, currently neither container can ping the other. If I start a third container just as test1 (with the env set) then test3 can ping test1 and vice versa but cannot see test2 (as expected.)

If I leave the env set and create test4 and/or test5 omitting -h or setting -h to just the new hostname (i.e. test5) respectively, the outcome is the same as test3.

What I really want is the host file generated as in test2. Is there some configuration I can apply to make that happen?

Thanks for your help!
-R

**Comment by rade** [comment]
> it seems that Ambari’s remote client does a reverse lookup after the initial, which it then pulls from the container’s local /etc/hosts.

I don't understand how Ambari gets the wrong answer here. If it does a reverse lookup of the unqualified name then the entry in /etc/hosts will return the correct IP. And if it does a reverse lookup of the fully qualified name then weaveDNS will return the correct IP.

What error do you get from Ambari?

> I didn’t set the environment before starting the container. Why does that control how /etc/hosts gets written back to the container?

Because if you don't set the environment then you are just starting a container normally via docker. No weave involved. And the IP address in /etc/hosts will be the docker network IP, not the weave network IP.

> What I really want is the host file generated as in test2. Is there some configuration I can apply to make that happen?

No. It's something we can add, but I am reluctant to do that w/o fully understanding the problem.

**Comment by rmharris157** [comment]
Understandible.

So, let's say I set the env var and fire off two docker containers, ambari server and a minimal ssh daemon running inside a centos:6 container. 

core@alpha ~ $ docker run -d -p 8080 --name amb-01 -h amb-01.weave.local sequenceiq/ambari
core@alpha ~ $ docker run -d --name hdp-01 -h hdp-01.weave.local leverege/ambari-drone
core@alpha ~ $ docker exec amb-01 ping -c 1 hdp-01.weave.local
PING hdp-01.weave.local (10.32.0.2) 56(84) bytes of data.
64 bytes from hdp-01.weave.local (10.32.0.2): icmp_seq=1 ttl=64 time=0.073 ms

--- hdp-01.weave.local ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.073/0.073/0.073/0.000 ms

So far, so good. I push all the keys around appropriately and login over the web to the Ambari cluster install wizard. I supply the key and hdp-01.weave.local as the host.

Ambari runs for a while and then fails with:

('ERROR 2015-10-02 13:14:31,889 main.py:146 - Ambari agent machine hostname (hdp-01) does not match expected ambari server hostname (hdp-01.weave.local). Aborting registration. Please check hostname, hostname -f and /etc/hosts file to confirm your hostname is setup correctly.

Hmmmm.... odd...
core@alpha ~ $ docker exec hdp-01 hostname
hdp-01.weave.local
core@alpha ~ $ docker exec hdp-01 hostname -f
hdp-01.weave.local
core@alpha ~ $ docker exec hdp-01 cat /etc/hosts
# modified by weave
10.32.0.2	hdp-01
127.0.0.1	localhost
::1	ip6-localhost ip6-loopback localhost
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters

So I can only speculate that somewhere ambari is resolving hdp-01 to the IP and then is doing a reverse IP lookup to get "hdp-01" back because I never refer to anything in the process as "hdp-01".

Thoughts?
-R

**Comment by rade** [comment]
side note:
> ```
core@alpha ~ $ docker run -d -p 8080 --name amb-01 -h amb-01.weave.local sequenceiq/ambari
core@alpha ~ $ docker run -d --name hdp-01 -h hdp-01.weave.local leverege/ambari-drone
```

you don't need the `-h` in any of the above since the hostname defaults to the container name + `.weave.local`.

> So I can only speculate that somewhere ambari is resolving hdp-01 to the IP and then is doing a reverse IP lookup to get "hdp-01" back.

Sounds plausible, though rather odd.

Anyway, since a) adding the FQDN does work, b) docker adds the FQDN too, and c) I cannot see any harm in adding the FQDN, we might as well do this.

**Comment by inercia** [comment]
Then, to make sure we are on the same page, the fix would be to add the hostname _and_ the FQDN in `/etc/hosts` (eg, `10.32.0.2   hdp-01.weave.local hdp-01`), right?.

Would we want this to be configurable (for example, with a `--with-fqdn-hostname` argument)?

**Comment by rmharris157** [comment]
Yes. 

I would assume that you would add both by default, unless you can think of a reason NOT to, at which point making the switch configurable would be fine too.
-R

**Comment by rade** [comment]
>  the fix would be to add the hostname *and* the FQDN

Correct.

> Would we want this to be configurable

Not separately; only as part of `--no-rewrite-hosts`.
